GPII-1783: Harden gpii-pouchdb for wider reuse...

the-t-in-rtf opened the pull request:
See [GPII-1783](https://issues.gpii.net/browse/GPII-1783) for details.

the-t-in-rtf committed:
GPII-1755:  Updated `gpii-express` version to pick up "error handler" middleware.  Also pushed through first `gpii-express` namespace change.

the-t-in-rtf committed:
GPII-1642:  Updated to latest version of `gpii.express`, including router/middleware changes and namespace updates.  Updated our own namespace as well to use the common conventions.

the-t-in-rtf committed:
GPII-1642:  Updated to latest `gpii-express` and `infusion`.  Also migrated all (two) tests to use `fluid.test.runTests`.

the-t-in-rtf committed:
GPII-1675:  Updated `gpii-express` to pick up diagnostic information in middleware wrappers.  Updated to newest `pouch` and `express-pouch` and improved isolation between runs.

the-t-in-rtf committed:
GPII-1642:  Updated `gpii-express` to pick up diagnostic information in router wrappers.

the-t-in-rtf committed:
GPII-1642:  Updated to pick up `gpii.express` routing table diagnostic function.

the-t-in-rtf committed:
GPII-1642:  Updated to latest `gpii.express` to pick up "non test safe" router diagnostics.

the-t-in-rtf committed:
GPII-1733:  Updated to pick up `error-capable` handlers from `gpii-express`.

the-t-in-rtf commented:
@amb26, this is "in progress", I shared it to start gathering some of your existing thoughts more "in context". You can already see the new "destroy" method in use as well as the latest version of PouchDB.

 I will update the ticket, comment here, and ping you once the work is ready for review.  In the meantime, please feel free to comment on anything else that needs work.

amb26 commented on src/js/pouch.js:
```diff
@@ -94,31 +95,17 @@ gpii.pouch.cleanup = function (that) {
-            promises.push(promise);
-        }
-    );
+        var promise = db.destroy()
```
Absolutely brilliant

the-t-in-rtf replied:
The context here scared the hell out of me, I thought I had omitted the semicolon like some kind of beast.

amb26 commented on tests/js/all-tests.js:
```diff
@@ -1,4 +1,9 @@
 "use strict";
+
+var events = require("events");
+events.EventEmitter.defaultMaxListeners = 100;
```
What requirement does this meet - that is, what complains if it is removed?

the-t-in-rtf replied:
It's a variation on the behavior described here:

http://www.jongleberry.com/understanding-possible-eventemitter-leaks.html

Node itself complains if you fail to clean up more than 10 listeners in a given run.

amb26 replied:
Yes - but what is the reason? Which event has more than 10 listeners added?

the-t-in-rtf replied:
We could not find a root cause in a brief exploration.  We have documented this and left it in place for now.

amb26 commented on tests/js/reload-tests.js:
```diff
@@ -65,16 +66,16 @@ fluid.defaults("gpii.pouch.tests.reload.caseHolder", {
-    gradeNames: ["gpii.pouch.tests.environment"],
+fluid.defaults("gpii.tests.pouch.reload.environment", {
+    gradeNames: ["gpii.test.pouch.environment"],
     port:       6792,
```
To aid confidence, could we supply this with the same port number (that is, the default) as the other environment?

the-t-in-rtf replied:
This has already been addressed in our call (by moving the old defaults to the test harness launcher).

amb26 commented on tests/js/environment.js:
```diff
@@ -1,6 +1,6 @@
 var fluid = require("infusion");
 
-fluid.defaults("gpii.pouch.tests.environment", {
+fluid.defaults("gpii.test.pouch.environment", {
```
Most of the rest of the feedback is low-priority, but it would be great if this environment could be properly factored so that I can eliminate the huge cut and paste I currently have in Kettle tests at
https://github.com/amb26/kettle/blob/KETTLE-32/tests/shared/gpii-pouchdb-harness.js
The largest issue is just hoisting the "databases" up somewhere (and fixing the processing of "data" would save a tiny workaround)

the-t-in-rtf replied:
I will look at it, but we should discuss in our chat tomorrow so I can make the update Friday morning.

the-t-in-rtf committed:
GPII-1783:  Updated to "self-deduping" `gpii-express`.

the-t-in-rtf committed:
GPII-1783: Updated to latest `gpii-express` to pick up test fixture refactoring there.  Updated to latest dev version of Infusion.

the-t-in-rtf committed:
GPII-1783:  Added not about our hiding of the orphaned listeners that result in warnings when running the tests.

the-t-in-rtf committed:
GPII-1783: Renamed variables to clarify their purpose.

the-t-in-rtf committed:
GPII-1783:  Moved port options to avoid having defaults in the harness grade itself.

the-t-in-rtf committed:
GPII-1783: Made `databases` settable from the environment level.

the-t-in-rtf committed:
GPII-1783:  Renamed `onStarted` to `onAllReady` and `constructServer` to `constructFixtures`.

the-t-in-rtf committed:
GPII-1783:  Renamed `onAllReady` to `onFixturesConstructed` per feedback from Antranig.

amb26 commented on tests/js/harness.js:
```diff
@@ -7,10 +7,15 @@ require("gpii-express");
     gradeNames: ["fluid.component"],
-    port:       6789,
-    baseUrl:    "http://localhost:6789/",
+    databases: {
```
Please remove your own databases from this grade and put the remainder into a reusable base grade in lib/

amb26 replied:
In addition there needs to be a gpii.pouch.loadTestingSupport() directive of the familiar form which loads these from the outside. As an additional courtesy this should in turn call gpii.express.loadTestingSupport().

amb26 replied:
See here for extent of the remaining duplication
https://github.com/amb26/kettle/blob/KETTLE-32/tests/DataSourcePouchDBTests.js#L22

the-t-in-rtf replied:
I've added `loadTestingSupport`.  Calling `gpii.express.loadTestingSupport` from that block as the only means of loading those grades unfortunately results in "assertion outside test case" messages.  We should review together.

amb26 commented on tests/js/environment.js:
```diff
@@ -1,23 +1,28 @@
+            events: {
+                onHarnessReady: "onHarnessReady"
+            }
+        }
```
An options distribution would be useful here to distribute the user's exact grade to the harness component - otherwise they will just write one themselves

amb26 replied:
The issue is that you have double-relayed all of the options out from pouch up to harness up to testEnvironment - with the exception of "databases" which is only relayed one step up to harness. To cut down the duplication I suggest you aggregate all of these into an options area named "pouchConfig" and just relay that in bulk to the testEnvironment

the-t-in-rtf replied:
I refactored to avoid the double-relaying, and to isolate the default settings from the environment and harness.

I did not use the `pouchConfig` name, which is already used within the pouch grade itself, and for a different purpose.  I will add the docs shortly to clarify the total range of options and we can discuss whether additional work is needed.

amb26 commented on tests/js/environment.js:
```diff
@@ -1,23 +1,28 @@
     gradeNames: ["fluid.test.testEnvironment"],
     port:       6792,
     baseUrl:    "http://localhost:6792/",
     testUrl:    "/sample/",
```
This probably doesn't belong here. Should pick a single place, either the harness or the environment, for static configuration of this kind - and factor it off for your own tests.

the-t-in-rtf replied:
The `testUrl` was cruft and has been removed.  I also upgrade `baseUrl` to use `fluid.stringTemplate` and the `port` option, and tidied up the usages throughout the package.

amb26 commented on tests/js/harness.js:
```diff
@@ -7,10 +7,15 @@ require("gpii-express");
 require("../../");
```
More intelligible is fluid.require("%gpii-pouchdb");

the-t-in-rtf replied:
Done.

the-t-in-rtf committed:
GPII-1783:  Rearranged for easier reuse per PR feedback.

the-t-in-rtf committed:
GPII-1783:  Created documentation for the main and test components.  Renamed various options to clarify their purpose.

the-t-in-rtf commented:
I just finished a docs spike.  While looking at the code I also removed now-unused functions from the previous cleanup method and renamed various options that were ambiguously named.

the-t-in-rtf committed:
GPII-1783:   Added requested `pouchConfig` wrapper to support passing a wider range of configuration options from environment->pouch.

amb26 commented on src/js/pouch.js:
```diff
@@ -94,60 +88,35 @@ gpii.pouch.cleanup = function (that) {
-    namespace:        "pouch", // Namespace to allow other routers to put themselves in the chain before or after us.
-    pouchConfigPath:  pouchConfigPath,
-    pouchConfig: {
+    gradeNames: ["fluid.modelComponent", "gpii.express.middleware"],
```
This component is unnecessarily a modelComponent

the-t-in-rtf replied:
Fixed.

amb26 commented on tests/js/launch-test-harness.js:
```diff
@@ -3,8 +3,20 @@
 var fluid = require("infusion");
```
These 3 files are not tests but test infrastructure which is to be used by others - I would put them in src/test

the-t-in-rtf replied:
Done.

amb26 commented on index.js:
```diff
@@ -1,5 +1,18 @@
+fluid.registerNamespace("gpii.pouch");
+
+gpii.pouch.loadTestingSupport = function () {
+    // require("gpii-express");
```
I can't reproduce any error from uncommenting these lines - can you confirm + paste?

the-t-in-rtf replied:
Previously, the errors were revealed by running `tests/js/launch-test-harness.js`, which would immediately complain about the lack of tests and exit.  This is no longer the case, and I have uncommented the lines.

the-t-in-rtf committed:
GPII-1783:  Moved reusable test infrastructure to `src/test` per PR feedback and updated all references.

the-t-in-rtf committed:
GPII-1783:  Removed unnecessary `modelComponent` grade in favor of `fluid.component`.

the-t-in-rtf committed:
GPII-1783:  Reenabled automatic call to `gpii.express.loadTestingSupport()` as part of `gpii.pouch.loadTestingSupport()`.
